tools/outline_proxy_controller/outline_controller_server.cpp

```diff
@@ -12,132 +12,149 @@
-      for (;;) {
-        for (;;) {
-          boost::asio::async_read_until(socket_, boost::asio::dynamic_buffer(buffer, 1024), "}",
-                                        yield);
-          std::cout << buffer << std::endl;
-          clientCommand.append(buffer);
-          buffer.clear();
-          if (isValidJson(clientCommand)) {
-            std::cout << "Valid JSON" << std::endl;
-            break;
-          }
-        }
-        auto [err, result, action] = runClientCommand(clientCommand);
-        // TODO: replace the following code with a json library to handle special characters
-        response << "{\"statusCode\": " << err
-                 << ",\"returnValue\": \"" << result << "\""
-                 << ",\"action\": \"" << action << "\"}" << std::endl;
-        boost::asio::async_write(
-            socket_, boost::asio::buffer(response.str(), response.str().length()), yield);
-        std::cout << "Wrote back (" << response.str() << ") to unix socket" << std::endl;
-        clientCommand.clear();
-        response.str(std::string());
-      }
-    } catch (std::exception& e) {
-      socket_.close();
-    }
-  });
-}
+// Minimum length of JSON input from app
+static const int kJsonInputMinLength = 10;
 
-bool session::isValidJson(std::string str) {
-  std::stringstream ss;
-  boost::property_tree::ptree pt;
+// The buffer size used to communicate with Outline client
+static const int kChannelBufferSize = 1024;
 
-  if (str.length() >= JSON_INPUT_MIN_LENGTH) {
-    ss << str;
+/**
+ * @brief Checks the input string and returns true if it's a valid json.
+ */
+static bool IsValidJson(const std::string &raw_str) {
+  if (raw_str.length() >= kJsonInputMinLength) {
     try {
-      boost::property_tree::read_json(ss, pt);
+      std::istringstream input{raw_str};
+      boost::property_tree::ptree pt;
+      boost::property_tree::read_json(input, pt);
       return true;
     } catch (std::exception const& e) {
     }
   }
   return false;
 }
 
-std::tuple<int, std::string, std::string> session::runClientCommand(std::string clientCommand) {
+OutlineClientSession::OutlineClientSession(
+  boost::asio::local::stream_protocol::socket &&channel,
+  std::shared_ptr<OutlineProxyController> outline_proxy_controller)
+  : channel_(std::move(channel)),
+    outline_controller_(outline_proxy_controller)
+{
+  logger.info("client session started");
+}
+
+OutlineClientSession::~OutlineClientSession() {
+  logger.info("client session terminated");
+}
+
+boost::asio::awaitable<void> OutlineClientSession::Start() {
+  using namespace boost::asio;
+
+  try {
+    std::string client_command, raw_buffer;
+    std::ostringstream response;
+    for (;;) {
+      do {
+        co_await async_read_until(
+          channel_, dynamic_buffer(raw_buffer, kChannelBufferSize), "}", use_awaitable);
+        client_command.append(raw_buffer);
+        raw_buffer.clear();
+      } while (!IsValidJson(client_command));
```

Comment:
Can you check the length here and clean up IsValidJson to not have that check?

```c
while (client_command.length() < kJsonInputMinLength || !IsValidJson(client_command)
```

Reply:
Yup, I was thinking of refactoring the IsValidJson to TryParseJson and pass the result ptree to RunClientCommand.

Reply:
Done.